docs/understanding-your-pipeline/canonical-event/index.md

```diff
@@ -36,274 +42,490 @@ The Snowplow data table is designed to be immutable: the data in each line shoul
+
+```sql
+with unique_events as (
+    select
+        ev.*,
+        row_number() over (partition by a.event_id order by a.collector_tstamp) as event_id_dedupe_index
+    from 
+        atomic.events ev
+),
+
+unique_my_custom_context as (
+    select
+        ctx.*,
+        row_number() over (partition by a.root_id order by a.root_tstamp) as my_custom_context_index
+    from 
+        atomic.my_custom_context_table ctx
+)
+
+select 
+    ...
+from 
+    unique_events u_ev
+left join 
+    unique_my_custom_context u_ctx
+    on u_ctx.root_id = u_ev.event_id and u_ctx.root_tstamp = u_ev.collector_tstamp and u_ctx.my_custom_context_index = 1
+where
+    u_ev.event_id_dedupe_index = 1
+```
+
+For the case of a potential multiple-record entity we need to employ a slightly more complex logic.
+
+First de-duplicate the events table in the same way as with a single entity context, but we also keep the number of duplicates there were. In the entity table we generate a row number per unique combination of **all** fields in the record. A join is then made on `root_id` and `root_tstamp` as before, but with an **additional** clause that the row number is a multiple of the number of duplicates, to support the 1-to-many join. This ensures all duplicates are removed while retaining all original records of the entity. This may look like a weird join condition, but it works.
+
+Unfortunately listing all fields manually can be quite tedious, but we have added support for this in the [de-duplication logic](/docs/modeling-your-data/modeling-your-data-with-dbt/dbt-advanced-usage/dbt-duplicates/index.md#multiple-entity-contexts) of our dbt packages.
+
+```sql
+with unique_events as (
+    select
+        ev.*,
+        row_number() over (partition by a.event_id order by a.collector_tstamp) as event_id_dedupe_index,
+    count(*) over (partition by a.event_id) as event_id_dedupe_count
+    from 
+        atomic.events ev
+),
+
+unique_my_custom_context as (
+    select
+        ctx.*,
+        row_number() over (partition by a.root_id, a.root_tstamp, ... /*all columns listed here for your context */ order by a.root_tstamp) as my_custom_context_index
+    from 
+        atomic.my_custom_context_table ctx
+)
+
+select 
+    ...
+from 
+    unique_events u_ev
+left join 
+    unique_my_custom_context u_ctx
+    on u_ctx.root_id = u_ev.event_id and u_ctx.root_tstamp = u_ev.collector_tstamp and mod(u_ctx.my_custom_context_index, u_ev.event_id_dedupe_count) 
+where
+    u_ev.event_id_dedupe_index = 1
+```
+
+</details>
+
+
+:::
+
+</TabItem>
+<TabItem value="bq" label="BigQuery">
+
+Contexts are not part of the standard `atomic.events` columns, it is instead a `REPATED` `RECORD` type column with the name of the schema and the full version number. The fields in this table will be determined by the JSON schema defined for the event in advance, converted to snake case. Users can query a single context's fields in the context by extracting them using 
```

Comment:
Nit: `REPEATED RECORD` over `REPEATED` `RECORD`

Reply:
I technically kept them separate as it is a `RECORD` type with the `REPEATED` property?